examples/chat2-reliable/reliability.go

```diff
@@ -0,0 +1,379 @@
+	var missingDeps []string
+	for _, depID := range msg.CausalHistory {
+		if !c.bloomFilter.Test(depID) {
+			missingDeps = append(missingDeps, depID)
+		}
+	}
+	return missingDeps
+}
+
+func (c *Chat) addToMessageHistory(msg *pb.Message) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	c.messageHistory = append(c.messageHistory, msg)
+	c.messageHistory = c.resolveConflicts(c.messageHistory)
+
+	if len(c.messageHistory) > maxMessageHistory {
+		c.messageHistory = c.messageHistory[len(c.messageHistory)-maxMessageHistory:]
+	}
+}
+
+func (c *Chat) resolveConflicts(messages []*pb.Message) []*pb.Message {
+	// Group messages by Lamport timestamp
+	groupedMessages := make(map[int32][]*pb.Message)
+	for _, msg := range messages {
+		groupedMessages[msg.LamportTimestamp] = append(groupedMessages[msg.LamportTimestamp], msg)
+	}
+
+	// Sort timestamps
+	var timestamps []int32
+	for ts := range groupedMessages {
+		timestamps = append(timestamps, ts)
+	}
+	sort.Slice(timestamps, func(i, j int) bool { return timestamps[i] < timestamps[j] })
+
+	// Resolve conflicts and create a new ordered list
+	var resolvedMessages []*pb.Message
+	for _, ts := range timestamps {
+		msgs := groupedMessages[ts]
+		if len(msgs) == 1 {
+			resolvedMessages = append(resolvedMessages, msgs[0])
+		} else {
+			// Sort conflicting messages by MessageId
+			sort.Slice(msgs, func(i, j int) bool { return msgs[i].MessageId < msgs[j].MessageId })
+			resolvedMessages = append(resolvedMessages, msgs...)
+		}
+	}
+
+	return resolvedMessages
+}
+
+func (c *Chat) periodicBufferSweep() {
+	defer c.wg.Done()
+
+	ticker := time.NewTicker(bufferSweepInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-c.ctx.Done():
+			return
+		case <-ticker.C:
+			// Process incoming buffer
+			c.processBufferedMessages()
+
+			// Resend unacknowledged messages from outgoing buffer
+			c.checkUnacknowledgedMessages()
+		}
+	}
+}
+
+func (c *Chat) checkUnacknowledgedMessages() {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	now := time.Now()
+	for i := 0; i < len(c.outgoingBuffer); i++ {
+		unackMsg := c.outgoingBuffer[i]
+		if now.Sub(unackMsg.SendTime) > ackTimeout {
+			if unackMsg.ResendAttempts < maxResendAttempts {
+				c.resendMessage(unackMsg.Message, unackMsg.ResendAttempts)
+				c.outgoingBuffer[i].ResendAttempts++
+				c.outgoingBuffer[i].SendTime = now
+			} else {
+				// Remove the message from the buffer after max attempts
+				c.outgoingBuffer = append(c.outgoingBuffer[:i], c.outgoingBuffer[i+1:]...)
+				i-- // Adjust index after removal
+				c.ui.ErrorMessage(fmt.Errorf("message %s dropped: failed to be acknowledged after %d attempts", unackMsg.Message.MessageId, maxResendAttempts))
+			}
+		}
+	}
+}
+
+func (c *Chat) resendMessage(msg *pb.Message, resendAttempts int) {
+	go func() {
+		delay := resendBaseDelay * time.Duration(1<<uint(resendAttempts))
+		if delay > maxResendDelay {
+			delay = maxResendDelay
+		}
+		time.Sleep(delay)
```

Comment:
Consider making this sleep 'cancellable':
```suggestion
select {
  case <-c.ctx.Done():
      return
  case <-time.After(delay):
     // do nothing
  }
```